### ISSUE TITLE
Provide a way to delete packaged component children at the end of `Construct` rather than at the end of the program

### ISSUE THREAD

**Issue by phillipedwards** [pulumi/pulumi#12636]
### What happened?

I have a MLC component that creates some resources and a program that creates the component. The component has a component method which generates a kubeconifg. The program calls this component method and passes the output result to an explicit kubernetes provider. The implementation of the method checks for the presence of certain resources in the cloud as part of its implementation and blocks before it returns.

Running a `pulumi up` for the first time works as expected. However, changing one of the component’s inputs which causes certain children to not be created (i.e. these should be deleted on next pulumi up). Running a pulumi up we expect those deletes to happen, but they don’t. This is due to the above mentioned component method blocking and waiting on the presence check never resolving as it is expecting those resources to be deleted, however, those resources do not get deleted until the end of the programs run.

### Expected Behavior

Deletes in a component should happen before the end of the program, thereby allowing this type of a component method to complete.

### Steps to reproduce

TBD

### Output of `pulumi about`

3.60.1

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by justinvp** [comment]
Deletes are performed at the end of a program. We may want to consider providing a way to perform any deletes of a component's children right after the `Construct` call, since we know the component is finished, rather than waiting for the end of the program.

However, it's possible that another part of the same program would want to define the same resource with an alias. It's also possible for a program to add children to any resource (outside of a component).

If we were to consider adding support for this, it might make sense as a resource option specified on the component implementation, or perhaps better, as an extra option on the `Construct` result, that you explicitly opt-in to to perform deletes after `Construct` for component children, with the caveat that if you’re later adding resources outside the component that are aliased or adding children outside of the component, that those would be out of scope.
